flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/BoundedBlockingSubpartitionReader.java

```diff
@@ -0,0 +1,191 @@
+
+	/** The reader/decoder to the memory mapped region with the data we currently read from.
+	 * Max 2GB large. Further regions may be in the {@link #furtherData} field.
+	 * Null once the reader empty or disposed.*/
+	@Nullable
+	private BoundedBlockingSubpartitionMemory.Reader data;
+
+	/** The next buffer (look ahead). Null once the data is depleted or reader is disposed. */
+	@Nullable
+	private Buffer nextBuffer;
+
+	/** The remaining number of data buffers (not events) in the result. */
+	private int bufferBacklog;
+
+	/** Flag whether this reader is released. Atomic, to avoid double release. */
+	private boolean isReleased;
+
+	/**
+	 * Convenience constructor that takes a single buffer.
+	 */
+	BoundedBlockingSubpartitionReader(
+			BoundedBlockingSubpartition parent,
+			ByteBuffer data,
+			int numBuffers) throws IOException {
+		this(parent, Collections.singleton(data), numBuffers);
+	}
+
+	/**
+	 * The general constructor that takes a queue of buffers containing the data.
+	 */
+	BoundedBlockingSubpartitionReader(
+			BoundedBlockingSubpartition parent,
+			Iterable<ByteBuffer> data,
+			int numBuffers) throws IOException {
+
+		checkArgument(numBuffers >= 0);
+
+		this.parent = checkNotNull(parent);
+		this.bufferBacklog = numBuffers;
+
+		this.furtherData = data.iterator();
+		checkArgument(furtherData.hasNext());
+
+		this.data = new BoundedBlockingSubpartitionMemory.Reader(furtherData.next());
+		this.nextBuffer = fillNextBuffer();
+	}
+
+	@Nullable
+	@Override
+	public BufferAndBacklog getNextBuffer() throws IOException {
+		final Buffer current = nextBuffer; // copy reference to stack
+
+		if (current == null) {
+			// as per contract, we must return null when the reader is empty,
+			// but also in case the reader is disposed (rather than throwing an exception)
+			return null;
+		}
+		if (current.isBuffer()) {
+			bufferBacklog--;
+		}
+
+		Buffer next = fillNextBuffer();
+		nextBuffer = next;
+
+		return new BufferAndBacklog(
+				current,
+				next != null,
+				bufferBacklog,
+				next != null && !next.isBuffer());
+	}
+
+	@Nullable
+	private Buffer fillNextBuffer() throws IOException {
+		// should only be null once empty or disposed, in which case this method
+		// should not be called any more
+		assert data != null;
+
+		final Buffer next = data.sliceNextBuffer();
+		if (next != null) {
+			return next;
+		}
+
+		if (!furtherData.hasNext()) {
+			data = null;
+			return null;
+		}
+
+		data = new BoundedBlockingSubpartitionMemory.Reader(furtherData.next());
+		return fillNextBuffer();
+	}
+
+	@Override
+	public void notifyDataAvailable() {
+		throw new IllegalStateException("No data should become available on a blocking partition during consumption.");
+	}
+
+	@Override
+	public void notifySubpartitionConsumed() throws IOException {
+		parent.onConsumedSubpartition();
+	}
+
+	@Override
+	public void releaseAllResources() throws IOException {
+		// mark released and prevent double release
+		if (isReleased) {
```

Comment:
We should add lock here to avoid mulitple thread entering when `isReleased` is false.

Reply:
The threading model here assumes only a singe thread as the user of the reader, so we do not need that lock.

All operations here are also idempotent, except the final `parent.releaseReaderReference()`, which is actually thread-safe, so this method can work without synchronization.